source/common/memory/utils.cc

```diff
@@ -18,22 +21,52 @@ void Utils::releaseFreeMemory() {
 
 /*
   The purpose of this function is to release the cache introduced by tcmalloc,
-  mainly in xDS config updates, admin handler, and so on. all work on the main thread,
+  mainly in xDS config updates, admin handler, and so on. All work on the main thread,
   so the overall impact on performance is small.
   Ref: https://github.com/envoyproxy/envoy/pull/9471#discussion_r363825985
 */
 void Utils::tryShrinkHeap() {
 #ifdef TCMALLOC
-  // TODO(zyfjeff): Make max unfreed memory byte configurable
-  static const uint64_t MAX_UNFREED_MEMORY_BYTE = 100 * 1024 * 1024;
+  uint64_t shrink_threshold = 100 * 1024 * 1024UL;
   auto total_physical_bytes = Stats::totalPhysicalBytes();
   auto allocated_size_by_app = Stats::totalCurrentlyAllocated();
+  Runtime::Loader* runtime = Runtime::LoaderSingleton::getExisting();
+  static Envoy::Stats::Counter* shrink_attempt_counter = nullptr;
+  static Envoy::Stats::Counter* free_memory_counter = nullptr;
+
+  if (runtime) {
+    shrink_threshold = runtime->threadsafeSnapshot()->getInteger(
+        "envoy.memory.heap_shrink_threshold", shrink_threshold);
+    if (shrink_attempt_counter) {
+      Envoy::Stats::Scope& root_scope = runtime->getRootScope();
+
+      Envoy::Stats::StatNameManagedStorage shrink_attempt_stat_name("envoy.memory_shrink_attempt",
+                                                                    root_scope.symbolTable());
+
+      shrink_attempt_counter = &root_scope.counterFromStatName(shrink_attempt_stat_name.statName());
```

Comment:
@jmarantz This is my attempt to save the lookup.
I save the ptr in functions' static var, but I am not sure if there is life time issue.

Any suggestion?

Reply:
I think you mean `!shink_attempt_counter` (same below). Will defer to @jmarantz on the best way to do this.

Reply:
fixed and and add the test
Not sure why the stats does't show up in test. Create another comment